src/runtime_src/core/pcie/driver/linux/xocl/subdev/flash.c

```diff
@@ -18,13 +19,1109 @@
+ */
+static int flash_rx(struct xocl_flash *flash, u8 *buf, size_t len)
+{
+	size_t cnt;
+
+	for (cnt = 0; cnt < len; cnt++) {
+		u8 c = flash_read8(flash);
+		if (buf)
+			buf[cnt] = c;
+	}
+
+	if ((flash_get_status(flash) & QSPI_SR_RX_EMPTY) == 0) {
+		FLASH_ERR(flash, "failed to drain RX fifo");
+		return -EINVAL;
+	}
+
+	if (flash_has_err(flash))
+		return -EINVAL;
+
+	return 0;
+}
+
+/*
+ * Caller should make sure the fifo is large enough to host len bytes.
+ */
+static int flash_tx(struct xocl_flash *flash, u8 *buf, size_t len)
+{
+	u32 ctrl = flash_get_ctrl(flash);
+	int i;
+
+	BUG_ON(len > flash->qspi_fifo_depth);
+
+	/* Stop transfering to the flash. */
+	flash_set_ctrl(flash, ctrl | QSPI_CR_TRANS_INHIBIT);
+
+	/* Fill out the FIFO. */
+	for (i = 0; i < len; i++)
+		flash_send8(flash, buf[i]);
+
+	/* Start transfering to the flash. */
+	flash_set_ctrl(flash, ctrl & ~QSPI_CR_TRANS_INHIBIT);
+
+	/* Waiting for FIFO to become empty again. */
+	if (FLASH_BUSY_WAIT(flash_get_status(flash) &
+		(QSPI_SR_TX_EMPTY | QSPI_SR_ERRS))) {
+		if (flash_has_err(flash)) {
+			FLASH_ERR(flash, "QSPI write failed");
+		} else {
+			FLASH_ERR(flash, "QSPI write timeout, status: 0x%x",
+				flash_get_status(flash));
+		}
+		return -ETIMEDOUT;
+	}
+
+	/* Always stop transfering to the flash after we finish. */
+	flash_set_ctrl(flash, ctrl | QSPI_CR_TRANS_INHIBIT);
+
+	if (flash_has_err(flash))
+		return -EINVAL;
+
+	return 0;
+}
+
+/*
+ * Reset both RX and TX FIFO.
+ */
+static int flash_reset_fifo(struct xocl_flash *flash)
+{
+	const u32 status_fifo_mask = QSPI_SR_TX_FULL | QSPI_SR_RX_FULL |
+		QSPI_SR_TX_EMPTY | QSPI_SR_RX_EMPTY;
+	u32 fifo_status = flash_get_status(flash) & status_fifo_mask;
+
+	if (fifo_status == (QSPI_SR_TX_EMPTY | QSPI_SR_RX_EMPTY))
+		return 0;
+
+	flash_set_ctrl(flash, flash_get_ctrl(flash) | QSPI_CR_TXFIFO_RESET |
+		QSPI_CR_RXFIFO_RESET);
+
+	if (FLASH_BUSY_WAIT((flash_get_status(flash) & status_fifo_mask)) ==
+		(QSPI_SR_TX_EMPTY | QSPI_SR_RX_EMPTY)) {
+		FLASH_ERR(flash, "failed to reset FIFO, status: 0x%x",
+			flash_get_status(flash));
+		return -ETIMEDOUT;
+	}
+	return 0;
+}
+
+static int flash_transaction(struct xocl_flash *flash,
+	u8 *buf, size_t len, int slave, bool need_output)
+{
+	int ret = 0;
+
+	/* Reset both the TX and RX fifo before starting transaction. */
+	ret = flash_reset_fifo(flash);
+	if (ret)
+		return ret;
+
+	/* The slave index should be within range. */
+	if (slave >= MAX_NUM_OF_SLAVES)
+		return -EINVAL;
+	flash_set_slave(flash, slave);
+
+	ret = flash_tx(flash, buf, len);
+	if (ret)
+		return ret;
+
+	if (need_output) {
+		ret = flash_rx(flash, buf, len);
+	} else {
+		/* Needs to drain the FIFO even when the data is not wanted. */
+		ret = flash_rx(flash, NULL, len);
+	}
+
+	/* Always need to reset slave select register after each transaction */
+	flash_set_slave(flash, SLAVE_NONE);
+
+	return 0;
+}
+
+static size_t flash_get_fifo_depth(struct xocl_flash *flash)
```

Comment:
The FIFO depth should be gotten from the DSA. The depth is configured when the DSA is built. Shouldn't be pushing junk into the flash controller to determine the depth.

Reply:
This is how xilinx's upstream spi driver finds out the fifo depth (see xilinx_spi_find_buffer_size() in https://elixir.bootlin.com/linux/latest/source/drivers/spi/spi-xilinx.c). So, I think this should be OK?

Reply:
That's an extremely ugly way of getting the depth of the buffer but I guess its fine if that's what they do in our spi drivers.

Reply:
yes, i agree that it'd be ideal if we have a register to read the fifo length instead of doing this. but that register is not available :-(.